src/43F0.c

```diff
@@ -74,9 +74,128 @@ HeapNode* _heap_create(s32* addr, u32 size) {
     }
 }
 
-INCLUDE_ASM(s32, "43F0", _heap_malloc);
+#define _heap_alloc_and_update_id(node) { \
+    u16 HeapEntryID = D_NEXT_HEAP_MALLOC_ID; \
+    node->allocated = 1; \
+    D_NEXT_HEAP_MALLOC_ID = HeapEntryID + 1; \
+    node->entryID = HeapEntryID; \
+}
+
+void* _heap_malloc(HeapNode* head, u32 size) {
+    HeapNode* nextHeapNode;
+    HeapNode* pPrevHeapNode = 0;
+    u32 newBlockSize;
+    u32 curBlockLength;
+    HeapNode* curHeapNode;
+    u32 smallestBlockFound;
+
+    //must allocate 16 bytes or more at minimum or fail
+    size = ALIGN16(size);
+    if (size == 0) {
+        return (s32 *)pPrevHeapNode;
+    }
+
+    smallestBlockFound = 0;
+    nextHeapNode = 0;
+
+    //find the smallest block we can fit into in the free list
+    for (curHeapNode = head;;curHeapNode = curHeapNode->next) {
+        if (curHeapNode->allocated == 0) {
+            curBlockLength = curHeapNode->length;
+            if ((curBlockLength >= size) && ((curBlockLength < smallestBlockFound) || (smallestBlockFound == 0))) {
+                pPrevHeapNode = curHeapNode;
+                smallestBlockFound = curBlockLength;
+                nextHeapNode = curHeapNode->next;
+            }
+        }
+        if (!curHeapNode->next) {
```

Comment:
I'd be surprised if this didn't match as the condition in the for loop

Reply:
Putting it as part of the loop condition results in an extra check early in the code to confirm the next pointer prior to doing the for loop.

Reply:
Perhaps it'd be a better match as `while`?

Reply:
do {} while(curHeapNode->next); crc fails

Reply:
I do agree it looks odd though